File: packages/typescript/src/pdf/vat-return.ts
// ─── VAT Return PDF (Layer 2) ────────────────────────────────────────────────
// Generates VAT return summary PDFs using pdfkit.
// Returns a Buffer, performs no file I/O. Serverless-compatible.

import PDFDocument from 'pdfkit';

// ─── Types ───────────────────────────────────────────────────────────────────

export interface VatReturnInvoice {
  invoiceNumber: string;
  amount: number;
  vatAmount: number;
  rateType: 'standard' | 'zero-rated' | 'exempt';
}

export interface VatReturnOptions {
  period: string; // YYYY-MM
  businessName: string;
  businessTin: string;
  businessVrn: string;
  salesInvoices: VatReturnInvoice[];
  purchaseInvoices?: VatReturnInvoice[];
}

export interface VatReturn {
  period: string;
  businessName: string;
  businessTin: string;
  businessVrn: string;
  outputVat: number;
  inputVat: number;
  netVatPayable: number;
  standardRatedSales: number;
  zeroRatedSales: number;
  exemptSales: number;
  invoiceCount: number;
  filingDeadline: string;
  salesInvoices: VatReturnInvoice[];
  purchaseInvoices: VatReturnInvoice[];
}


// ─── Formatting Helpers ──────────────────────────────────────────────────────

function formatCurrency(amount: number): string {
  const fixed = Math.abs(amount).toFixed(2);
  const [whole, decimal] = fixed.split('.');
  const withCommas = whole.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  const sign = amount < 0 ? '-' : '';
  return `${sign}\u20A6${withCommas}.${decimal}`;
}

function formatDate(iso: string): string {
  const parts = iso.split('-');
  if (parts.length !== 3) return iso;
  return `${parts[2]}/${parts[1]}/${parts[0]}`;
}

function bankersRound2(value: number): number {
  const shifted = value * 100;
  const floored = Math.floor(shifted);
  const remainder = shifted - floored;
  const epsilon = 1e-9;
  if (Math.abs(remainder - 0.5) < epsilon) {
    return (floored % 2 === 0 ? floored : floored + 1) / 100;
  }
  return Math.round(shifted) / 100;
}

/**
 * Compute the filing deadline: 21st of the month following the period.
 */
function computeFilingDeadline(period: string): string {
  const [yearStr, monthStr] = period.split('-');
  let year = parseInt(yearStr, 10);
  let month = parseInt(monthStr, 10) + 1;
  if (month > 12) {
    month = 1;
    year += 1;
  }
  return `${year}-${String(month).padStart(2, '0')}-21`;
}

// ─── Layout Constants ────────────────────────────────────────────────────────

const PAGE_MARGIN = 50;
const PAGE_WIDTH = 595.28;
const CONTENT_WIDTH = PAGE_WIDTH - PAGE_MARGIN * 2;
const FONT_BOLD = 'Helvetica-Bold';
const FONT_REGULAR = 'Helvetica';
const COLOR_PRIMARY = '#1a1a2e';
const COLOR_SECONDARY = '#555555';
const COLOR_LINE = '#cccccc';

// ─── Generate ────────────────────────────────────────────────────────────────

/**
 * Generate a VatReturn object from options.
 * Computes output VAT, input VAT, net VAT payable, breakdown by rate type,
 * invoice count, and filing deadline.
 */
export function generate(options: VatReturnOptions): VatReturn {
  const purchaseInvoices = options.purchaseInvoices ?? [];

  let outputVat = 0;
  let standardRatedSales = 0;
  let zeroRatedSales = 0;
  let exemptSales = 0;

  for (const inv of options.salesInvoices) {
    outputVat = bankersRound2(outputVat + inv.vatAmount);
    if (inv.rateType === 'standard') {
      standardRatedSales = bankersRound2(standardRatedSales + inv.amount);
    } else if (inv.rateType === 'zero-rated') {
      zeroRatedSales = bankersRound2(zeroRatedSales + inv.amount);
    } else {
      exemptSales = bankersRound2(exemptSales + inv.amount);
    }
  }

  let inputVat = 0;
  for (const inv of purchaseInvoices) {
    inputVat = bankersRound2(inputVat + inv.vatAmount);
  }

  const netVatPayable = bankersRound2(outputVat - inputVat);
  const invoiceCount = options.salesInvoices.length + purchaseInvoices.length;
  const filingDeadline = computeFilingDeadline(options.period);

  return {
    period: options.period,
    businessName: options.businessName,
    businessTin: options.businessTin,
    businessVrn: options.businessVrn,
    outputVat,
    inputVat,
    netVatPayable,
    standardRatedSales,
    zeroRatedSales,
    exemptSales,
    invoiceCount,
    filingDeadline,
    salesInvoices: options.salesInvoices,
    purchaseInvoices,
  };
}

// ─── PDF Generation ──────────────────────────────────────────────────────────

/**
 * Generate a VAT Return summary PDF as a Buffer.
 * Renders: business details, period, output VAT, input VAT, net VAT payable,
 * breakdown by rate type, invoice count, filing deadline.
 */
export function toPDF(ret: VatReturn): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({
      size: 'A4',
      margin: PAGE_MARGIN,
      info: {
        Title: `VAT Return - ${ret.businessName} - ${ret.period}`,
        Author: ret.businessName,
      },
    });

    const chunks: Uint8Array[] = [];
    doc.on('data', (chunk: Uint8Array) => chunks.push(chunk));
    doc.on('end', () => resolve(Buffer.concat(chunks)));
    doc.on('error', (err: Error) => reject(err));

    // ── Title ──
    doc.font(FONT_BOLD).fontSize(14).fillColor(COLOR_PRIMARY)
      .text('VAT RETURN SUMMARY', PAGE_MARGIN, PAGE_MARGIN, {
        align: 'center', width: CONTENT_WIDTH,
      });

    doc.moveDown(0.5);
    doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY)
      .text(`Period: ${ret.period}`, { align: 'center', width: CONTENT_WIDTH });
    doc.text(`Filing Deadline: ${formatDate(ret.filingDeadline)}`, { align: 'center', width: CONTENT_WIDTH });

    doc.moveDown(1);

    // ── Business Details ──
    doc.font(FONT_BOLD).fontSize(10).fillColor(COLOR_PRIMARY).text('Business Details');
    doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY);
    doc.text(`Name: ${ret.businessName}`);
    doc.text(`TIN: ${ret.businessTin}`);
    doc.text(`VRN: ${ret.businessVrn}`);

    doc.moveDown(1);

    // ── VAT Summary ──
    doc.font(FONT_BOLD).fontSize(10).fillColor(COLOR_PRIMARY).text('VAT Summary');
    doc.moveDown(0.3);

    const labelX = PAGE_MARGIN + 10;
    const valueX = PAGE_MARGIN + 280;

    const summaryRows: [string, string][] = [
      ['Output VAT (on sales):', formatCurrency(ret.outputVat)],
      ['Input VAT (on purchases):', formatCurrency(ret.inputVat)],
      ['Net VAT Payable:', formatCurrency(ret.netVatPayable)],
    ];

    doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY);
    for (const [label, value] of summaryRows) {
      const y = doc.y;
      doc.text(label, labelX, y);
      doc.font(FONT_BOLD).text(value, valueX, y, { width: 180, align: 'right' });
      doc.font(FONT_REGULAR);
      doc.y = y + 16;
    }

    doc.moveDown(1);

    // ── Breakdown by Rate Type ──
    doc.font(FONT_BOLD).fontSize(10).fillColor(COLOR_PRIMARY).text('Sales Breakdown by Rate Type');
    doc.moveDown(0.3);

    const breakdownRows: [string, string][] = [
      ['Standard-rated sales:', formatCurrency(ret.standardRatedSales)],
      ['Zero-rated sales:', formatCurrency(ret.zeroRatedSales)],
      ['Exempt sales:', formatCurrency(ret.exemptSales)],
      ['Total invoices:', String(ret.invoiceCount)],
    ];

    doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY);
    for (const [label, value] of breakdownRows) {
      const y = doc.y;
      doc.text(label, labelX, y);
      doc.text(value, valueX, y, { width: 180, align: 'right' });
      doc.y = y + 16;
    }

    doc.moveDown(2);

    // ── Footer ──
    doc.moveTo(PAGE_MARGIN, doc.y).lineTo(PAGE_MARGIN + CONTENT_WIDTH, doc.y)
      .strokeColor(COLOR_LINE).lineWidth(0.5).stroke();
    doc.y += 8;

    doc.font(FONT_REGULAR).fontSize(7).fillColor(COLOR_SECONDARY)
      .text(
        'This VAT return summary is generated in compliance with the Nigeria Tax Act (NTA) 2025 and VAT Act as amended. ' +
        'VAT returns must be filed with the Nigeria Revenue Service (NRS) by the 21st of the month following the taxable period.',
        PAGE_MARGIN, doc.y, { width: CONTENT_WIDTH },
      );

    doc.end();
  });
}

// ─── NRS Format ──────────────────────────────────────────────────────────────

/**
 * Convert a VatReturn to NRS-compatible filing format.
 * Returns a plain object suitable for JSON serialization.
 */
export function toNRSFormat(ret: VatReturn): object {
  return {
    formType: 'VAT_RETURN',
    period: ret.period,
    filingDeadline: ret.filingDeadline,
    business: {
      name: ret.businessName,
      tin: ret.businessTin,
      vrn: ret.businessVrn,
    },
    outputVat: ret.outputVat,
    inputVat: ret.inputVat,
    netVatPayable: ret.netVatPayable,
    salesBreakdown: {
      standardRated: ret.standardRatedSales,
      zeroRated: ret.zeroRatedSales,
      exempt: ret.exemptSales,
    },
    invoiceCount: ret.invoiceCount,
    salesInvoices: ret.salesInvoices.map((inv) => ({
      invoiceNumber: inv.invoiceNumber,
      amount: inv.amount,
      vatAmount: inv.vatAmount,
      rateType: inv.rateType,
    })),
    purchaseInvoices: ret.purchaseInvoices.map((inv) => ({
      invoiceNumber: inv.invoiceNumber,
      amount: inv.amount,
      vatAmount: inv.vatAmount,
      rateType: inv.rateType,
    })),
  };
}

File: packages/python/src/ngtaxkit/pdf/vat_return.py
"""VAT Return Summary PDF generation using fpdf2."""

from __future__ import annotations

from typing import Any

from . import _require_fpdf2
from .helpers import (
    COLOR_HEADER_BG, COLOR_LINE, COLOR_PRIMARY, COLOR_SECONDARY, COLOR_WHITE,
    CONTENT_WIDTH, PAGE_MARGIN, format_currency, format_date,
)


def _cur(amount: float) -> str:
    """Format currency for PDF output (latin-1 safe)."""
    return format_currency(amount, pdf_safe=True)


def create(
    *,
    period: str,
    business_name: str,
    business_tin: str,
    business_vrn: str,
    sales_invoices: list[dict[str, Any]],
    purchase_invoices: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    """Build a VAT Return data dict with computed totals."""
    output_vat = sum(inv.get("vat_amount", 0) for inv in sales_invoices)
    input_vat = sum(
        inv.get("vat_amount", 0) for inv in (purchase_invoices or [])
    )
    net_vat = output_vat - input_vat

    return {
        "period": period,
        "business_name": business_name,
        "business_tin": business_tin,
        "business_vrn": business_vrn,
        "sales_invoices": sales_invoices,
        "purchase_invoices": purchase_invoices or [],
        "output_vat": output_vat,
        "input_vat": input_vat,
        "net_vat": net_vat,
        "sales_count": len(sales_invoices),
        "purchase_count": len(purchase_invoices or []),
    }


def to_pdf(vat_return: dict[str, Any]) -> bytes:
    """Generate a VAT Return Summary PDF as bytes."""
    fpdf = _require_fpdf2()
    pdf = fpdf.FPDF(orientation="P", unit="mm", format="A4")
    pdf.set_auto_page_break(auto=True, margin=20)
    pdf.add_page()

    # -- VAT RETURN SUMMARY banner --
    pdf.set_fill_color(*COLOR_PRIMARY)
    pdf.set_text_color(*COLOR_WHITE)
    pdf.set_font("Helvetica", "B", 14)
    pdf.cell(CONTENT_WIDTH, 10, "VAT RETURN SUMMARY", fill=True, align="C",
             new_x="LMARGIN", new_y="NEXT")
    pdf.ln(4)

    # -- Business details --
    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.cell(0, 4, f"Business: {vat_return['business_name']}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"TIN: {vat_return['business_tin']}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"VRN: {vat_return['business_vrn']}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"Period: {vat_return['period']}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.ln(6)

    # -- Output VAT section --
    total_sales = sum(
        inv.get("amount", 0) for inv in vat_return["sales_invoices"]
    )
    pdf.set_font("Helvetica", "B", 10)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.cell(0, 5, "Output VAT (Sales)", new_x="LMARGIN", new_y="NEXT")
    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.cell(0, 4,
             f"Total Sales: {_cur(total_sales)}  |  Output VAT: {_cur(vat_return['output_vat'])}  |  Invoices: {vat_return['sales_count']}",
             new_x="LMARGIN", new_y="NEXT")
    pdf.ln(4)

    # -- Input VAT section (if purchase invoices) --
    if vat_return["purchase_invoices"]:
        total_purchases = sum(
            inv.get("amount", 0) for inv in vat_return["purchase_invoices"]
        )
        pdf.set_font("Helvetica", "B", 10)
        pdf.set_text_color(*COLOR_PRIMARY)
        pdf.cell(0, 5, "Input VAT (Purchases)",
                 new_x="LMARGIN", new_y="NEXT")
        pdf.set_font("Helvetica", "", 9)
        pdf.set_text_color(*COLOR_SECONDARY)
        pdf.cell(0, 4,
                 f"Total Purchases: {_cur(total_purchases)}  |  Input VAT: {_cur(vat_return['input_vat'])}  |  Invoices: {vat_return['purchase_count']}",
                 new_x="LMARGIN", new_y="NEXT")
        pdf.ln(4)

    # -- Net VAT Payable highlighted --
    pdf.ln(2)
    col_label = 110
    col_value = 70
    pdf.set_fill_color(*COLOR_PRIMARY)
    pdf.set_text_color(*COLOR_WHITE)
    pdf.set_font("Helvetica", "B", 12)
    pdf.cell(col_label, 10, "  NET VAT PAYABLE", fill=True, align="L")
    pdf.cell(col_value, 10, f"{_cur(vat_return['net_vat'])}  ", fill=True,
             align="R")
    pdf.ln()
    pdf.ln(6)

    # -- Sales invoice listing table --
    pdf.set_font("Helvetica", "B", 10)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.cell(0, 5, "Sales Invoices", new_x="LMARGIN", new_y="NEXT")
    pdf.ln(1)

    s_col_widths = [45, 50, 45, 40]
    s_headers = ["Invoice #", "Amount", "VAT", "Rate Type"]

    pdf.set_fill_color(*COLOR_HEADER_BG)
    pdf.set_font("Helvetica", "B", 8)
    pdf.set_text_color(*COLOR_PRIMARY)
    for i, h in enumerate(s_headers):
        align = "L" if i == 0 or i == 3 else "R"
        pdf.cell(s_col_widths[i], 6, h, border=0, fill=True, align=align)
    pdf.ln()

    pdf.set_font("Helvetica", "", 8)
    pdf.set_text_color(*COLOR_SECONDARY)
    for inv in vat_return["sales_invoices"]:
        pdf.cell(s_col_widths[0], 5,
                 str(inv.get("invoice_number", ""))[:25], align="L")
        pdf.cell(s_col_widths[1], 5,
                 _cur(inv.get("amount", 0)), align="R")
        pdf.cell(s_col_widths[2], 5,
                 _cur(inv.get("vat_amount", 0)), align="R")
        pdf.cell(s_col_widths[3], 5,
                 str(inv.get("rate_type", "")).replace("-", " ").title(),
                 align="L")
        pdf.ln()
        pdf.set_draw_color(*COLOR_LINE)
        pdf.line(pdf.l_margin, pdf.get_y(),
                 pdf.l_margin + CONTENT_WIDTH, pdf.get_y())
        pdf.ln(1)

    pdf.ln(4)

    # -- Purchase invoice listing table (if any) --
    if vat_return["purchase_invoices"]:
        pdf.set_font("Helvetica", "B", 10)
        pdf.set_text_color(*COLOR_PRIMARY)
        pdf.cell(0, 5, "Purchase Invoices",
                 new_x="LMARGIN", new_y="NEXT")
        pdf.ln(1)

        pdf.set_fill_color(*COLOR_HEADER_BG)
        pdf.set_font("Helvetica", "B", 8)
        pdf.set_text_color(*COLOR_PRIMARY)
        for i, h in enumerate(s_headers):
            align = "L" if i == 0 or i == 3 else "R"
            pdf.cell(s_col_widths[i], 6, h, border=0, fill=True, align=align)
        pdf.ln()

        pdf.set_font("Helvetica", "", 8)
        pdf.set_text_color(*COLOR_SECONDARY)
        for inv in vat_return["purchase_invoices"]:
            pdf.cell(s_col_widths[0], 5,
                     str(inv.get("invoice_number", ""))[:25], align="L")
            pdf.cell(s_col_widths[1], 5,
                     _cur(inv.get("amount", 0)), align="R")
            pdf.cell(s_col_widths[2], 5,
                     _cur(inv.get("vat_amount", 0)), align="R")
            pdf.cell(s_col_widths[3], 5,
                     str(inv.get("rate_type", "")).replace("-", " ").title(),
                     align="L")
            pdf.ln()
            pdf.set_draw_color(*COLOR_LINE)
            pdf.line(pdf.l_margin, pdf.get_y(),
                     pdf.l_margin + CONTENT_WIDTH, pdf.get_y())
            pdf.ln(1)

        pdf.ln(4)

    # -- Legal footer --
    pdf.set_draw_color(*COLOR_LINE)
    pdf.line(pdf.l_margin, pdf.get_y(),
             pdf.l_margin + CONTENT_WIDTH, pdf.get_y())
    pdf.ln(3)
    pdf.set_font("Helvetica", "", 7)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.multi_cell(CONTENT_WIDTH, 3,
                   "This VAT return summary is prepared in accordance with the Nigeria Tax Act (NTA) 2025 "
                   "and the Value Added Tax Act as amended. VAT is charged at the applicable rate per NTA 2025 "
                   "Schedule 1 (zero-rated), Schedule 2 (exempt), or the standard rate of 7.5%.")
    pdf.ln(2)
    pdf.cell(CONTENT_WIDTH, 3,
             f"Generated by ngtaxkit -- VAT Return | {vat_return['period']} | {vat_return['business_tin']}",
             align="C")

    return pdf.output()
